src/libpriv/rpmostree-origin.cxx

```diff
@@ -772,9 +772,13 @@ rpmostree_origin_add_overrides (RpmOstreeOrigin  *origin,
        * later on in the core. This is just an early easy sanity check. */
       if (g_hash_table_contains (origin->cached_overrides_remove, pkg) ||
           g_hash_table_contains (origin->cached_overrides_local_replace, pkg))
-        return glnx_throw (error, "Override already exists for package '%s'", pkg);
+        {
+          g_hash_table_insert (origin->cached_overrides_local_replace, g_strdup (pkg),
+                             util::move_nullify (sha256));
+          g_hash_table_add (origin->cached_overrides_remove, g_strdup (pkg));
```

Comment:
I'm not very familiar with this logic, but this seems suspicious. Shouldn't any existing entry be removed from the `cached_overrides_remove` map?

Additionally, I believe that this has further implications in the `if (changed)` block further on, as there is now an action which touches both maps at the same time. Possibly both would need to be updated in keyfile.

Reply:
Thanks for the review.
I see your point about needing to remove the entry, I just thought it was solved by this characteristic of the function g_hash_table_add () https://developer.gnome.org/glib/stable/glib-Hash-Tables.html#g-hash-table-add : 
```
In particular, this means that if key already exists in the hash table, 
then the old copy of key in the hash table is freed and key replaces it in the table.
```

I'm going to look into how to do the testcases and also the changes implications in the other block.

Reply:
We shouldn't add to both here. Packages in `cached_overrides_local_replace` represent base packages we're replacing. Packages in `cached_overrides_remove` represent base packages we're removing. In this function, we should be adding to only one of the two (either we're replacing a base package, or we're removing it -- which one is requested depends on the `type` argument). The check here just verifies that we're not trying to duplicate an existing override, so I think this check should remain.  (I've mentioned this elsewhere, but probably these should've been idempotent from the start so that we just ignored already existing and matching overrides, but we can't change that default behaviour now.)

So conceptually, the flow we're interested in supporting here is to be able to swap out an entry in `cached_overrides_local_replace` for another. Concretely, this ends up being two separate operations: removing one `NEVRA1 -> SHA256` entry and adding back another `NEVRA2 -> HA256` entry.

So the tricky bit here is that we want to replace based on the package name, but we store NEVRAs. So to handle this, we'd have to split the NEVRA of the passed package, and split the NEVRA of all the override replacements to find a match, which is cumbersome.

Instead, what I would suggest is to do this at the next level up in `deploy_transaction_execute`. We have this related piece of code there: https://github.com/coreos/rpm-ostree/blob/00cd0878bf8e22a7baa9691b83b57cf68cb78eb2/src/daemon/rpmostreed-transaction-types.cxx#L1203-L1278

This code builds a mapping of name <--> NEVRA. So what we can do is rework this code so that when we have replacements, we check if there's already an existing replacement for that package name, and if so, call `rpmostree_origin_remove_override(..., RPMOSTREE_ORIGIN_OVERRIDE_REPLACE_LOCAL)` on it with the matching NEVRA1 before we call `rpmostree_origin_add_overrides(..., RPMOSTREE_ORIGIN_OVERRIDE_REPLACE_LOCAL)` for the new `NEVRA2:SHA256`.